### ISSUE TITLE
[Security] add 'is_granted' to default security ExpressionLanguage?

### ISSUE THREAD

**Issue by dmaicher** [symfony/symfony#23084]
| Q                | A
| ---------------- | -----
| Bug report?      | no
| Feature request? | yes
| BC Break report? | no
| RFC?             | no
| Symfony version  | 3.4

In one of my projects I would like to evaluate a complex expression involving checking other custom attribute voters.

Unfortunately this does not work out of the box:

```php
$authChecker->isGranted(new Expression("is_granted('custom_attr') and/or ..."))
```

==> `The function "is_granted" does not exist around position 1 for expression ...`

My solution for now is to add a custom extension for the security ExpressionLanguage and tag it with `security.expression_language_provider`:

```php
use Symfony\Component\DependencyInjection\ContainerInterface;
use Symfony\Component\ExpressionLanguage\ExpressionFunction;
use Symfony\Component\ExpressionLanguage\ExpressionFunctionProviderInterface;

class ExpressionLanguageProvider implements ExpressionFunctionProviderInterface
{
    /**
     * @var ContainerInterface
     */
    private $container;

    public function __construct(ContainerInterface $container)
    {
        $this->container = $container;
    }

    public function getFunctions()
    {
        return [
            new ExpressionFunction('is_granted', null, function (array $variables, $attributes, $object = null) {
                return $this->container->get('security.authorization_checker')->isGranted($attributes, $object);
            })
        ];
    }
}
```

As @stof pointed out here https://github.com/symfony/symfony-docs/pull/4282 there is a circular dependency and that's why I get the `security.authorization_checker` directly from the container. It would also be possible to use a `ServiceLocator ` for it in 3.4.

WDYT? Should we think about adding this to the core? To me it seems quite useful.

**Comment by Hanmac** [comment]
i think it's nice, but i don't know if you need the whole ContainerInterface when you can just use a `AuthorizationCheckerInterface`? Edit: oh okay i see with "circular dependency" why it doesn't work.

hm i don't know the deeper part, but can it be directly added to `Symfony\Component\Security\Core\Authorization\ExpressionLanguageProvider`? Hm i think it can't because it doesn't work with `AuthenticationTrustResolverInterface`?

need to think whats the better way would be to add this into core

**Comment by derrabus** [comment]
I tried to use `is_granted` in an ACL expression a couple of weeks ago and was disappointed that it did not work as expected. I've managed to work around it, yet I would see a benefit in adding this feature.

**Comment by dmaicher** [comment]
If we include it in the core it will make this obsolete though: https://github.com/dmaicher/SensioFrameworkExtraBundle/blob/3.0/Security/ExpressionLanguage.php#L27

**Comment by dmaicher** [comment]
@derrabus good point :blush: This would of course mean we can also use `is_granted(...)` on ACL expressions :+1:

**Comment by jvasseur** [comment]
The circular dependency should be resolved in Symfony 3.3 now that voters are lazy loaded.

**Comment by dmaicher** [comment]
@jvasseur ah good point :+1: I'm still on 2.8 with this particular app where I'm using it :yum:
